### ISSUE TITLE
There is no .joernIndex folder

### ISSUE THREAD

**Issue by st152840** [SySeVR/SySeVR#11]
I'm not able to find .joernIndex directory. Whats the reason of the problem. I'm running the source code using joern.jar file. But I dont see any output generated. And also my database gets no update.

**Comment by sanashahir** [comment]
Pls do ctrl+h in the folder of joern...and change the location of neo4j database as path of .joernIndex folder

**Comment by st152840** [comment]
Thanks for your comments. Now it works I can see the database. I have changed my config path to .joernIndex and I tried ruuning db , it works. But still when I run the get_cfg_relation.py file, I still get 404 error in connecting to the database.

**Comment by sanashahir** [comment]
How do you start the database..did you close the browser?? And how do you run the python file..??

**Comment by st152840** [comment]
Thank you for you fast reply. I'm starting the data base using joern-start-db file. And I have changed the config file to .joernIndex. And then I'm starting the python script using python2..

py2neo.error.GraphError: HTTP GET returned response 404

**Comment by st152840** [comment]
File "test.py", line 9, in <module>
    j.connectToDatabase()
  File "/usr/local/lib/python2.7/dist-packages/joern/all.py", line 33, in connectToDatabase
    self.gremlin = Gremlin(self.graphDb)

**Comment by sanashahir** [comment]
I started the db using the command sudo ./neo4j start-no-wait in the terminal.. and then sudo python2 get*******.py

**Comment by st152840** [comment]
Ya I also tried the same but still I get the database error. But one thing I also want to add is I'm using neo4j-4.0.4 version. Will it make any problem?. Which version are you using?

**Comment by sanashahir** [comment]
I am using neo4j 2.1.8 joern 0.3.1 supports only this version

**Comment by st152840** [comment]
oh is it..I 'm not able to find the exact version. It would be great if you could share some link or resources regarding it.. Thanks

**Comment by sanashahir** [comment]
[](https://github.com/tophertimzen/vagrantized-Joern/blob/master/setup.sh)  please look on this link

**Comment by sanashahir** [comment]
[](https://github.com/tophertimzen/vagrantized-Joern/blob/master/setup.sh)  please look on this link

**Comment by st152840** [comment]
Thanks a lot... I'll try and let you know.. :)

**Comment by sanashahir** [comment]
Okay

**Comment by st152840** [comment]
Thanks a lot. Now it works successfully and I'm able to see the graphs also in neo4j browser. Now I'm trying to do further. Thanks a lot for your inputs. Are you still working in this project?

**Comment by sanashahir** [comment]
Okay good!! You are welcome!! Yah I am doing this project. I got stuck in prediction..

**Comment by st152840** [comment]
great. Are you also doing for the same CVEs'? or some different CVE?

**Comment by sanashahir** [comment]
Yeah the CVEs' given in the program data folder. And What about the project process?

**Comment by st152840** [comment]
I'm just in the data_preprocessing....I had some other works in between so couldn't do a lot...But I think we need to change a lot in the code to make it function...There are many index out of range problems...What do you think about that?

**Comment by sanashahir** [comment]
Yeah you are right,  there are so many errors, labelling is not in right way. Need to correct data_preprocess.py

**Comment by st152840** [comment]
so you have fixed those things?

**Comment by sanashahir** [comment]
errors that occur has changed in my own way.but i am not getting any idea to find solution for data_preprocess.py for labelling the slices. And do you have any idea about to do prediction from the text file??

**Comment by st152840** [comment]
I'm also just trying to debug this code...I'll update if I have any positive results....You can also let me know if you get some positive results...

**Comment by st152840** [comment]
I'm currently working on process_dataflow_func.py....The previous step was data_preprocess.py. There i was able to generate two text files with some labels. But there were some errors. I have attached the generated labels file.
[pointersuse_slices.txt](https://github.com/SySeVR/SySeVR/files/4756944/pointersuse_slices.txt)
[api_slices.txt](https://github.com/SySeVR/SySeVR/files/4756945/api_slices.txt)

**Comment by st152840** [comment]
can you check the following code for make_label. I did it only with two CWE.

`## coding:utf-8

import os

import pickle





f = open("dict_cwe2father.pkl", 'rb')

dict_cwe2father = pickle.load(f)

f.close()



#print dict_cwe2father['CWE-787']



f = open("label_vec_type.pkl", 'rb')

label_vec_type = pickle.load(f)

f.close()



f = open("dict_testcase2code.pkl",'rb')

dict_testcase2code = pickle.load(f)

f.close()





def get_label_veclist(list_cwe):

    list_label = [0] * len(label_vec_type)

    for cweid in list_cwe:

        index = label_vec_type.index(cweid)

        list_label[index] = 1



    return list_label





def get_label_cwe(cweid, label_cwe):

    if cweid in label_vec_type:

        label_cwe.append(cweid)

        return label_cwe



    else:

        if cweid == 'CWE-1000':

            label_cwe = label_vec_type

        else:

            fathercweid = dict_cwe2father[cweid]



            for _id in fathercweid:

                label_cwe = get_label_cwe(_id, label_cwe)



    return label_cwe





def make_label(path, dict_vuln2testcase, _type):

    f = open(path, 'r')

    context = f.read().split('------------------------------')[:-1]

    f.close()



    context[0] = '\n' + context[0]



    list_all_label = []

    list_all_vulline = []

    for _slice in context:

        vulline = []

        index_line = _slice.split('\n')[1] 

        list_codes = _slice.split('\n')[2:-1] 

        case_name = index_line.split(' ')[1]

        key_name = '/'.join(index_line.split(' ')[1].split('/')[-2:])

        print index_line



        if key_name in dict_vuln2testcase.keys():

            list_codeline = [code.split(' ')[-1] for code in list_codes]

            dict = dict_vuln2testcase[key_name]



            _dict_cwe2line_target = {}

            _dict_cwe2line = {}

	    for _dict in dict: 

                for key in _dict.keys():

                    if _dict[key] not in _dict_cwe2line_target.keys():

                        _dict_cwe2line_target[_dict[key]] = [key] 

                    else:

                        _dict_cwe2line_target[_dict[key]].append(key)



                

                for line in list_codeline:

                    line = line.strip()

                    if line in _dict.keys():

                        if not ' '.join((list_codes[list_codeline.index(line)].strip()).split(' ')[:-1]) == dict_testcase2code[key_name+"/"+line].strip():

                            continue

                        cweid = _dict[line]

                        vulline.append(list_codeline.index(line))



                        if cweid not in _dict_cwe2line.keys():

                            _dict_cwe2line[cweid] = [line]

                        else:

                            _dict_cwe2line[cweid].append(line)



            if _type:

                list_vuln_cwe = []

                for key in _dict_cwe2line.keys():

                    if key == 'Any...':

                        continue

                    if len(_dict_cwe2line[key]) == len(_dict_cwe2line_target[key]):

                        label_cwe = []

                        label_cwe = get_label_cwe(key, label_cwe)

                        list_vuln_cwe += label_cwe



            else:

                list_vuln_cwe = []

                for key in _dict_cwe2line.keys():

                    if key == 'Any...':

                        continue

                    label_cwe = []

                    label_cwe = get_label_cwe(key, label_cwe)

                    list_vuln_cwe += label_cwe



            if list_vuln_cwe == []:

                list_label = [0] * len(label_vec_type)

            else:

                list_vuln_cwe = list(set(list_vuln_cwe))

                list_label = get_label_veclist(list_vuln_cwe)

            

        else:

            list_label = [0] * len(label_vec_type)



        list_all_label.append(list_label)

        list_all_vulline.append(vulline)



    return list_all_label, list_all_vulline





def main():

    f = open("dict_flawline2filepath.pkl", 'rb')

    dict_vuln2testcase = pickle.load(f)

    f.close()

    _type = False

    time = '4'

    lang = '/sysevr/SySeVR/Implementation/source2slice/test_data/' + time

    

    path = os.path.join(lang, 'api_slices.txt')

    #path = os.path.join('api_slices.txt')

    list_all_apilabel, list_all_vulline = make_label(path, dict_vuln2testcase, _type)

    dec_path = os.path.join(lang, 'api_slices.pkl')

    f = open(dec_path, 'wb')

    pickle.dump(list_all_apilabel, f, True)

    f.close()

    dec_path = os.path.join(lang, 'api_slices_vulline.pkl')

    f = open(dec_path, 'wb')

    pickle.dump(list_all_vulline, f)

    f.close()

    

    # path = os.path.join(lang, 'arraysuse_slices.txt')

    # list_all_arraylabel = make_label(path, dict_vuln2testcase, _type)

    # dec_path = os.path.join(lang, 'array_slice_label.pkl')

    # f = open(dec_path, 'wb')

    # pickle.dump(list_all_arraylabel, f, True)

    # f.close()

    

    path = os.path.join(lang, 'pointersuse_slices.txt')

    list_all_pointerlabel = make_label(path, dict_vuln2testcase, _type)

    dec_path = os.path.join(lang, 'pointersuse_slices.pkl')

    f = open(dec_path, 'wb')

    pickle.dump(list_all_pointerlabel, f, True)

    f.close()

 

    # path = os.path.join(lang, 'integeroverflow_slices.txt')

    # list_all_exprlabel = make_label(path, dict_vuln2testcase, _type)

    # dec_path = os.path.join(lang, 'expr_slice_label.pkl')

    # f = open(dec_path, 'wb')

    # pickle.dump(list_all_exprlabel, f, True)

    # f.close()

    



if __name__ == '__main__':

    main()

`

**Comment by sanashahir** [comment]
Okay let me check it...if you don't mind please give me ur email...I have some other doubts

**Comment by st152840** [comment]
my e-mail id is jisaachenderson@gmail.com

**Comment by sanashahir** [comment]
> my e-mail id is [jisaachenderson@gmail.com](mailto:jisaachenderson@gmail.com)

OK thanks, I have mailed some doubts of mine regarding the project

**Comment by QiuJYWX** [comment]
Hi @sanashahir @st152840 , I have got a problem when running 'pyhon2 extract_df.py'. I used the NVD dataset, totally there are 874 subfolders in the NVD dataset, 874 subfolders in the cfg_db folder, 874subfolders in the pdg_db folder, and 846 subfolders in the dict_call2cfgNodeID_funcID folders. It can be seen that some information of specific functions was not successfully extracted. Thus the extract_df.py got an error with:

File "extract_df.py", line 479, in
api_slice()
File "extract_df.py", line 349, in api_slice
list_code, startline, startline_path = program_slice(pdg, list_sensitive_funcid, slice_dir, key)
File "extract_df.py", line 302, in program_slice
list_cross_func_back, not_scan_func_list = process_crossfuncs_back_byfirstnode(start_list, testID, i, not_scan_func_list)
File "/home/qjy/Desktop/SySeVR/Implementation/source2slice/slice_op.py", line 349, in process_crossfuncs_back_byfirstnode
fin = open(path, 'rb')
IOError: [Errno 2] No such file or directory: u'dict_call2cfgNodeID_funcID/CVE-2015-4517/dict.pkl'

Anyone can help me about this issue?

**Comment by QiuJYWX** [comment]
Hi @st152840 @sanashahir ,
Thanks for your time and attention. I have got the following error when running extract_df.py:
Traceback (most recent call last):
File "extract_df.py", line 479, in
api_slice()
File "extract_df.py", line 349, in api_slice
list_code, startline, startline_path = program_slice(pdg, list_sensitive_funcid, slice_dir, key)
File "extract_df.py", line 302, in program_slice
list_cross_func_back, not_scan_func_list = process_crossfuncs_back_byfirstnode(start_list, testID, i, not_scan_func_list)
File "/home/qjy/Desktop/SySeVR/Implementation/source2slice/slice_op.py", line 362, in process_crossfuncs_back_byfirstnode
if func_name not in _dict.keys():
UnboundLocalError: local variable '_dict' referenced before assignment

Anyone can provide me some advice?

**Comment by st152840** [comment]
hey @QiuJYWX 

Regarding your first question, ya true there might be some corrupted files while processing the graphs. So just try deleting them and try it out.


Regarding 2nd one I have never faced that error...But one thing is very important when you don't use the code paths which they mentioned you'll get lot of errors. If your using your own paths to store some files or so..Then you have to make some changes in the code.

**Comment by QiuJYWX** [comment]
> hey @QiuJYWX
> 
> Regarding your first question, ya true there might be some corrupted files while processing the graphs. So just try deleting them and try it out.
> 
> Regarding 2nd one I have never faced that error...But one thing is very important when you don't use the code paths which they mentioned you'll get lot of errors. If your using your own paths to store some files or so..Then you have to make some changes in the code.

Hi @st152840 ,
Thanks for your kindly reply. I will debug the project according to your suggestions. 
BTW, did you successfully run this project?

**Comment by st152840** [comment]
ya i ran it..but in the final model.py there were lot of changes which we have to do. I just tried for one of the four vulnerabilities.

**Comment by QiuJYWX** [comment]
> ya i ran it..but in the final model.py there were lot of changes which we have to do. I just tried for one of the four vulnerabilitie

Excellent! I will ask you for suggestion when I encounter new issues.

**Comment by QiuJYWX** [comment]
> > ya i ran it..but in the final model.py there were lot of changes which we have to do. I just tried for one of the four vulnerabilitie
> 
> Excellent! I will ask you for suggestion when I encounter new issues.

Hi @st152840,

Thanks for your time and attention. I tried to run the bgru.py and encounted several issues in the metrics part. Would you please share you code script or give me some suggestions?

Regards,

**Comment by hiwamn1** [comment]
> Pls do ctrl+h in the folder of joern...and change the location of neo4j database as path of .joernIndex folder

Hello, 

I just executed joern.jar and I also could not find the .joernIndex file. Which file should I change to see this file? I have access to a server using ssh and I don't have a graphical interface. Could you please explain more?
